### PR TITLE
Forward_explain & Export with XAI: Detection & InstSeg

### PR DESCRIPTION
**Pull request by GalyaZalesskaya**
### Summary

- Detection & Instance Segmentation
- [x] Forward_explain 
- [x] Export with XAI
- [x] Tests
- [x] Validation on 
    - [x] PyToch task
    - [x] OV task 
- [x] Feature vectors implementation

There are several issues with XAI integration tests, so some of them were disabled or skipped to be further investigated.

### Checklist

- [x] I have added unit tests to cover my changes.​
- [x] I have added integration tests to cover my changes.​
- [x] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by negvet — DISMISSED** [review]

**Comment by GalyaZalesskaya** [comment]
@sungmanc @eunwoosh @vinnamkim @jaegukhyun Can you please review this PR?

The coverage tests are failing but I'd like to add them in another PR, since 
- this PR is too big already
- although `forward_explain` is not covered by unit tests, it's covered by integration tests, so the functionality is being tested.
Thank you :)

File: src/otx/core/exporter/mmdeploy.py [hunk 1]
```diff
@@ -63,14 +63,19 @@ def __init__(
         swap_rgb: bool = False,
         metadata: dict[tuple[str, str], str] | None = None,
         max_num_detections: int = 0,
+        output_names: list[str] | None = None,
```
Comment (eunwoosh) [inline src/otx/core/exporter/mmdeploy.py:67]: Could you add docstring for this?

File: src/otx/core/model/entity/detection.py [hunk 2]
```diff
@@ -105,42 +107,111 @@ def _export_parameters(self) -> dict[str, Any]:
...
+            self._customize_outputs(outputs, inputs)
+            if self._customize_outputs != ExplainableOTXDetModel._customize_outputs
+            else outputs["predictions"]
+        )
+
+    @staticmethod
+    def _forward_explain_detection(
+        self: SingleStageDetector,
+        inputs: torch.Tensor,
+        data_samples: OptSampleList = None,
```
Comment (eunwoosh) [inline src/otx/core/model/entity/detection.py]:
```suggestion
        data_samples: OptSampleList | None = None,
```

File: tests/integration/cli/test_cli.py [hunk 3]
```diff
@@ -208,8 +208,11 @@ def test_otx_e2e(
     assert latest_dir.exists()
 
     # 5) otx export with XAI
-    if "_cls" not in task or "dino" in model_name:
-        return
+    if ("_cls" not in task) and (task not in ["detection", "instance_segmentation"]):
+        pytest.skip("Supported only for classification, detection and instance segmentation task.")
```
Comment (eunwoosh) [inline tests/integration/cli/test_cli.py]:
```suggestion
        return  # Supported only for classification, detection and instance segmentation task.
```
If pytest.skip exists here, model, which doesn't support XAI, is considered as 'skip' even if it passes everything except XAI.

File: tests/integration/cli/test_cli.py [hunk 4]
```diff
@@ -208,8 +208,11 @@ def test_otx_e2e(
...
+
+    if "dino" in model_name or "rtmdet_inst_tiny" in model_name:
+        pytest.skip("DINO and Rtmdet_tiny are not supported.")
```
Comment (eunwoosh) [inline tests/integration/cli/test_cli.py]:
```suggestion
        return  # DINO and Rtmdet_tiny are not supported.
```
Same as above

**Review by eunwoosh — COMMENTED** [review]
Thanks for your work. I left some comments. Please take a look.

File: src/otx/core/exporter/mmdeploy.py (hunk 1 above)
Comment (GalyaZalesskaya) [reply src/otx/core/exporter/mmdeploy.py:67]: Fixed

File: tests/integration/cli/test_cli.py (hunk 3 above)
Comment (GalyaZalesskaya) [reply tests/integration/cli/test_cli.py]: Thank you for the comment, fixed

**Review by negvet — DISMISSED** [review]

**Review by negvet — APPROVED** [review]

**Review by vinnamkim — APPROVED** [review]

**Review by eunwoosh — APPROVED** [review]

**Pull request merged**